### PR TITLE
RES 1228: Refactor management of layer parameters

### PR DESCRIPTION
**Pull request by mvacaporale**
This pull request makes changes to the way parameters are managed, generally and specifically targeted for use in `resnets.py`.

Changes include:
* Separate param functionality  for conv, linear, and activation layers.
* `auto_sparse_parms` is now split into `auto_sparse_conv_params` and `auto_sparse_activation_params`
* `LayerParams` is now an abstract class which handles parameter specification for conv, linear, and activation layers through either a default dictionary or a function with arbitrary inputs.
* `LayerParamsByKeys` is introduced for easier management of parameters in specific use cases -`SpareWeightsLayerParams` is one specific use case.
* Both `resnets.py` and its tests have been refactored to conform to these changes.

### CODE REVIEW

File: nupic/research/frameworks/pytorch/sparse_layer_params.py [hunk 1]
```diff
@@ -19,37 +19,13 @@
...
+def auto_sparse_conv_params(in_channels, out_channels, kernel_size):
     """
-    Given conv2d parameters, automatically calculate sparsity parameters.
+    Given conv2d parameters, automatically calculate sparsity parameters
+    for `SparseWeights2d` module.
+
+    Note:
     This is highly experimental and likely to change.
 
     :return: an instance of LayerParams
```
Comment (subutai) [inline nupic/research/frameworks/pytorch/sparse_layer_params.py]: The return value should be changed.

File: nupic/research/frameworks/pytorch/sparse_layer_params.py [hunk 2]
```diff
@@ -76,19 +52,288 @@ def auto_sparse_params(in_channels, out_channels, kernel_size):
...
+        conv_source = self.conv_params_func or self.default_conv_params
+        act_source = self.activation_params_func or self.default_activation_params
+        return self.__class__.__name__ + \
+            "(\n\tlinear_params={},\n\tconv_params={},\n\tactivation_params={})".format(
+                repr(lin_source), repr(conv_source), repr(act_source)
+            )
+
+    def get_linear_params(self, *args, **params):
+
+        if self.linear_params_func:
```
Comment (subutai) [inline nupic/research/frameworks/pytorch/sparse_layer_params.py]: These should all be `if ... is not None`

File: nupic/research/frameworks/pytorch/sparse_layer_params.py [hunk 3]
```diff
@@ -76,19 +52,288 @@ def auto_sparse_params(in_channels, out_channels, kernel_size):
...
+                "activation_sparsity",  # <key name to pass to OtherParams.__init__>
+                ...,                    # `...` => No Default
+                "percent_on"            # "<target name for saving>"
+            )
+        ]
+
+    p = OtherParams(conv_sparsity=0.7)
+    print(p)
+    >>> OtherParams(
+        linear_params={"linear_sparisty": 0.5},
```
Comment (subutai) [inline nupic/research/frameworks/pytorch/sparse_layer_params.py]: `sparsity`

File: tests/unit/frameworks/pytorch/resnets_test.py [hunk 4]
```diff
@@ -26,15 +26,15 @@
...
-    cf_dict,
-    default_sparse_params,
     resnet18,
     resnet50,
     resnet101,
 )
-from nupic.research.frameworks.pytorch.sparse_layer_params import NoactLayerParams
+from nupic.research.frameworks.pytorch.sparse_layer_params import (
+    LayerParams,
+    SpareWeightsLayerParams,
```
Comment (subutai) [inline tests/unit/frameworks/pytorch/resnets_test.py]: Shouldn't this be `SparseWeightsLayerParams`?

Comment (mvacaporale) [reply tests/unit/frameworks/pytorch/resnets_test.py]: Fixed

Comment (mvacaporale) [reply tests/unit/frameworks/pytorch/resnets_test.py]: as well as your other comments preceding.

**Pull request merged**